content/en/_index.md

```diff
@@ -148,6 +181,14 @@ public static void exampleSnippet(String projectId, String filePath) {
     // Snippet content ...
 }
 {{< /tab >}}
+{{< tab header="Go" >}}
+// exampleSnippet demonstrates best practices.
+// The function prints to an io.Writer for testing purposes. Do not print to
+// stdout or stderr. Use the io.Writer instead.
+func exampleSnippet(w io.Writer, projectId, filePath string) {
```

Comment:
So for Java, we set and capture the equivalent of `fmt.Println` and just reset it once the test is over. Could we standardize Go doing that as well? 

The reason the og working group decided to try to limit/ban return values/errors was because it was a common pattern to just return something with no context. While it's easier to test, it's harder to run and understand how to interact with the response for the user.

Reply:
I agree the writer pattern isn't great and would prefer to just return values. I also understand that can produce unintuitive samples, too.

We're moving into hacky territory by trying to capture `fmt.Println`, but we could do something like this:

``` golang
package main

import (
	"fmt"
	"log"
	"os"
)

func init() {
	f, err := os.OpenFile("test.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0755)
	if err != nil {
		log.Fatal(err)
	}
	os.Stdout = f
}

func main() {
	fmt.Println("Hello there")
}
```

Reply:
See https://stackoverflow.com/a/10476304/927514 for example.

Reply:
In this approach, we'd be able to remove the io.Writer param, which would be quite nice.

Reply:
IMO, using a PIPE to a buffer is clearer, and this can be implemented on the test side (without showing it in the sample).

Reply:
I do think that passing a writer is better than overriding per test. Doing this is racy and means you can't run tests in parallel

Reply:
That seems to be the simplest approach, albeit strange by Go standards.

Reply:
Do we run our tests in parallel now?

Reply:
Yes, we do have samples that test in parallel today.

Reply:
Should we use logger instead? Then we can use an instance per sample.

Reply:
io.Writer might be nicer simply because it doesn't support things like Fatalf, etc.

Reply:
Consolidating this again on https://github.com/GoogleCloudPlatform/samples-style-guide/pull/43#discussion_r838999192.